go/spec/spec.go

```diff
@@ -0,0 +1,295 @@
+// constructed by parsing strings, either through ForDatabase, ForDataset, or
+// ForPath (or their Opts variations).
+type Spec struct {
+	// Spec is the spec string this was parsed into.
+	Spec string
+
+	// Protocol is one of "mem", "ldb", "http", or "https".
+	Protocol string
+
+	// DatabaseName is the name of the Spec's database, which is the string after
+	// "protocol:". http/https specs include their leading "//" characters.
+	DatabaseName string
+
+	// Options are the SpecOptions that the Spec was constructed with.
+	Options SpecOptions
+
+	// DatasetName is empty unless the spec was created with ForDataset.
+	DatasetName string
+
+	// Path is nil unless the spec was created with ForPath.
+	Path AbsolutePath
+
+	// db is lazily created, so it needs to be a pointer to a Database.
+	db *datas.Database
+}
+
+func newSpec(spec string, dbSpec string, opts SpecOptions) (Spec, error) {
+	protocol, dbName, err := parseDatabaseSpec(dbSpec)
+	if err != nil {
+		return Spec{}, err
+	}
+
+	return Spec{
+		Spec:         spec,
+		Protocol:     protocol,
+		DatabaseName: dbName,
+		Options:      opts,
+		db:           new(datas.Database),
+	}, nil
+}
+
+// ForDatabase parses a spec for a Database.
+func ForDatabase(spec string) (Spec, error) {
+	return ForDatabaseOpts(spec, SpecOptions{})
+}
+
+// ForDatabaseOpts parses a spec for a Database.
+func ForDatabaseOpts(spec string, opts SpecOptions) (Spec, error) {
+	return newSpec(spec, spec, opts)
+}
+
+// ForDataset parses a spec for a Dataset.
+func ForDataset(spec string) (Spec, error) {
+	return ForDatasetOpts(spec, SpecOptions{})
+}
+
+// ForDatasetOpts parses a spec for a Dataset.
+func ForDatasetOpts(spec string, opts SpecOptions) (Spec, error) {
+	dbSpec, dsName, err := splitDatabaseSpec(spec)
+	if err != nil {
+		return Spec{}, err
+	}
+
+	if !datasetRe.MatchString(dsName) {
+		return Spec{}, fmt.Errorf("Dataset %s must match %s", dsName, datasetRe.String())
+	}
+
+	sp, err := newSpec(spec, dbSpec, opts)
+	if err != nil {
+		return Spec{}, err
+	}
+
+	sp.DatasetName = dsName
+	return sp, nil
+}
+
+// ForPath parses a spec for a path to a Value.
+func ForPath(spec string) (Spec, error) {
+	return ForPathOpts(spec, SpecOptions{})
+}
+
+// ForPathOpts parses a spec for a path to a Value.
+func ForPathOpts(spec string, opts SpecOptions) (Spec, error) {
+	dbSpec, pathStr, err := splitDatabaseSpec(spec)
+	if err != nil {
+		return Spec{}, err
+	}
+
+	path, err := NewAbsolutePath(pathStr)
+	if err != nil {
+		return Spec{}, err
+	}
+
+	sp, err := newSpec(spec, dbSpec, opts)
+	if err != nil {
+		return Spec{}, err
+	}
+
+	sp.Path = path
+	return sp, nil
+}
+
+// GetDatabase returns the Database instance that this Spec's DatabaseName
+// describes. The same Database instance is returned every time, unless Close
+// is called. If the Spec is closed, it is re-opened with a new Database.
+func (sp Spec) GetDatabase() datas.Database {
+	if *sp.db == nil {
```

Comment:
I think you meant `sp.db` here. `*sp.db` can also be `nil` because `*sp.db` is of type `Database`, which is interface. But I'm not sure what it would mean for the pointer to be OK, but the db to be a nil interface.